Add support for suppressing clang static analyzer issues with [[clang::suppress]]

ddkilzer opened the pull request:
#### 22343e69f00f3f26fc5df4016a301a654b1c0bc3
<pre>
Add support for suppressing clang static analyzer issues with [[clang::suppress]]
<a href="https://bugs.webkit.org/show_bug.cgi?id=267981">https://bugs.webkit.org/show_bug.cgi?id=267981</a>
&lt;<a href="https://rdar.apple.com/121489134">rdar://121489134</a>&gt;

Reviewed by Alex Christensen and Darin Adler.

This adds support for the `[[clang::suppress]]` attribute in clang-17.

Currently there is no way to ignore specific analyzer warnings, but the
macros add support for documenting the warning(s) in case support is
added later.

The `[[clang::suppress]]` attribute can be applied to a single line of
code or to a block of code.

* Source/WTF/wtf/Compiler.h:
(COMPILER_HAS_ATTRIBUTE): Add.
- Define helper macro since __has_attribute() is used more than once.
(COMPILER_HAS_CLANG_FEATURE):
- Update documentation link to https.
(FALLTHROUGH):
(NOT_TAIL_CALLED):
- Use COMPILER_HAS_ATTRIBUTE().
(IGNORE_CLANG_STATIC_ANALYZER_WARNINGS_ATTRIBUTE): Add.
(IGNORE_CLANG_STATIC_ANALYZER_WARNINGS_BEGIN): Add.
(IGNORE_CLANG_STATIC_ANALYZER_WARNINGS_END): Add.
- Add support for the [[clang::suppress]] attribute.
(IGNORE_CLANG_STATIC_ANALYZER_USE_AFTER_MOVE_ATTRIBUTE): Add.
(TLS_MODEL_INITIAL_EXEC):
- Use COMPILER_HAS_ATTRIBUTE().

* Tools/TestWebKitAPI/Tests/WTF/CheckedPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/CompactPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/CompactRefPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/FixedVector.cpp:
* Tools/TestWebKitAPI/Tests/WTF/ListHashSet.cpp:
* Tools/TestWebKitAPI/Tests/WTF/NakedPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/PackedRefPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/RefPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/URL.cpp:
* Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/cf/RetainPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/cocoa/TypeCastsCocoa.mm:
* Tools/TestWebKitAPI/Tests/WTF/ns/RetainPtr.mm:
* Tools/TestWebKitAPI/Tests/WebKit/cocoa/WeakObjCPtr.mm:
- Ignore expected cplusplus.Move warnings using
  IGNORE_CLANG_STATIC_ANALYZER_USE_AFTER_MOVE_ATTRIBUTE.

Canonical link: <a href="https://commits.webkit.org/273460@main">https://commits.webkit.org/273460@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4bd35de58f17a0062ed0acce0b76d276867c827

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35506 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14452 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37643 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38248 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32015 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36704 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16833 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11485 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36058 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12209 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31618 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10711 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/31760 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39494 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/30106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32271 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32076 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/35389 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10920 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/8826 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34757 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/12641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/42063 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11437 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/8732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/11704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -564,6 +568,17 @@
+/* IGNORE_CLANG_STATIC_ANALYZER_WARNINGS_BEGIN() - whether the compiler supports suppressing static analysis warnings. */
+/* https://clang.llvm.org/docs/AttributeReference.html#suppress */
+#if COMPILER_HAS_CLANG_ATTRIBUTE(suppress)
+#define IGNORE_CLANG_STATIC_ANALYZER_WARNINGS_ATTRIBUTE(warning, ...) [[clang::suppress]]
```
We need this macro because other compilers will warn?

ddkilzer replied:
No.  This attribute will let us ignore clang static analyzer issues in specific lines/blocks of code so that we can start marking certain checkers as "clean" so that EWS can eventually tell us when new issues (for a given checker) are introduced.

However, this attribute isn't supported until clang-17.

Also note that per feedback from Alex, `COMPILER_HAS_CLANG_ATTRIBUTE()` has been renamed to `COMPILER_HAS_ATTRIBUTE()` since gcc also supports `__has_attribute()`.

achristensen07 commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -41,8 +41,16 @@
+/* COMPILER_HAS_CLANG_ATTRIBUTE() - whether the compiler supports a particular attribute. */
+/* https://clang.llvm.org/docs/LanguageExtensions.html#has-attribute */
+#ifdef __has_attribute
+#define COMPILER_HAS_CLANG_ATTRIBUTE(x) __has_attribute(x)
```
This should be just COMPILER_HAS_ATTRIBUTE since it doesn't have anything to do with clang here.
Also, can we use __has_cpp_attribute instead?  That's documented here: https://en.cppreference.com/w/cpp/feature_test#Attributes

ddkilzer replied:
> This should be just COMPILER_HAS_ATTRIBUTE since it doesn't have anything to do with clang here.

It appears that `__has_attribute()` is clang-specific, just like `__has_feature()`.

> Also, can we use __has_cpp_attribute instead? That's documented here: https://en.cppreference.com/w/cpp/feature_test#Attributes

 No, `__has_cpp_attribute()` won't work (change the test case to see it won't work):  <https://godbolt.org/z/nrehshT3o>

achristensen07 replied:
You're right __has_cpp_attribute does not work in this case.

achristensen07 reviewed (commented):
I feel like I was just reading about a way to mark a type as being able to be used safely after it is moved from.  Let me look and see if I can find that again.

darinadler commented on Tools/TestWebKitAPI/Tests/WTF/CheckedPtr.cpp:
```diff
@@ -110,6 +110,7 @@ TEST(WTF_CheckedPtr, Basic)
         EXPECT_EQ(checkedObject.ptrCount(), 2u);
         EXPECT_EQ(ptr1.get(), &checkedObject);
         EXPECT_EQ(ptr2.get(), &checkedObject);
+        IGNORE_CLANG_STATIC_ANALYZER_WARNINGS_ATTRIBUTE("cplusplus.Move")
```
If all uses at this time are going to be `"cplusplus.Move"`, then I suggest a specific macro for that one. We can add other specific macros for others that are used repeatedly.

ddkilzer replied:
The reason all uses are "cplusplus.Move" at this time is that I simply chose that checker to "fix" first.

Okay, I'll add a specific macro for that check.

Commit:
Add support for suppressing clang static analyzer issues with [[clang::suppress]]
https://bugs.webkit.org/show_bug.cgi?id=267981
<rdar://121489134>

Reviewed by Alex Christensen and Darin Adler.

This adds support for the `[[clang::suppress]]` attribute in clang-17.

Currently there is no way to ignore specific analyzer warnings, but the
macros add support for documenting the warning(s) in case support is
added later.

The `[[clang::suppress]]` attribute can be applied to a single line of
code or to a block of code.

* Source/WTF/wtf/Compiler.h:
(COMPILER_HAS_ATTRIBUTE): Add.
- Define helper macro since __has_attribute() is used more than once.
(COMPILER_HAS_CLANG_FEATURE):
- Update documentation link to https.
(FALLTHROUGH):
(NOT_TAIL_CALLED):
- Use COMPILER_HAS_ATTRIBUTE().
(IGNORE_CLANG_STATIC_ANALYZER_WARNINGS_ATTRIBUTE): Add.
(IGNORE_CLANG_STATIC_ANALYZER_WARNINGS_BEGIN): Add.
(IGNORE_CLANG_STATIC_ANALYZER_WARNINGS_END): Add.
- Add support for the [[clang::suppress]] attribute.
(IGNORE_CLANG_STATIC_ANALYZER_USE_AFTER_MOVE_ATTRIBUTE): Add.
(TLS_MODEL_INITIAL_EXEC):
- Use COMPILER_HAS_ATTRIBUTE().

* Tools/TestWebKitAPI/Tests/WTF/CheckedPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/CompactPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/CompactRefPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/FixedVector.cpp:
* Tools/TestWebKitAPI/Tests/WTF/ListHashSet.cpp:
* Tools/TestWebKitAPI/Tests/WTF/NakedPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/PackedRefPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/RefPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/URL.cpp:
* Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/cf/RetainPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/cocoa/TypeCastsCocoa.mm:
* Tools/TestWebKitAPI/Tests/WTF/ns/RetainPtr.mm:
* Tools/TestWebKitAPI/Tests/WebKit/cocoa/WeakObjCPtr.mm:
- Ignore expected cplusplus.Move warnings using
  IGNORE_CLANG_STATIC_ANALYZER_USE_AFTER_MOVE_ATTRIBUTE.

Canonical link: https://commits.webkit.org/273460@main
